Add ERC 8049: Contract-Level Onchain Metadata

nxt3d opened the pull request:
### Summary

Introduces a standard for storing contract-level metadata onchain using ERC-8042's Diamond Storage pattern. Extends ERC-7572's concept with onchain storage for cross-chain compatibility and upgradable contracts.

### Key Features

- **Diamond Storage**: Uses ERC-8042 for predictable storage locations
- **Cross-Chain Compatible**: Storage slots remain consistent across deployments  
- **ERC-7572 Compatible**: Maintains compatibility with existing contract metadata standards
- **Upgradable Support**: Metadata persists through contract upgrades

### Technical Details

- Diamond Storage pattern for predictable storage locations
- Support for name, description, image, etc.

### Example Use Case

**Contract ENS Naming**: Any contract can store its ENS name using this standard:
- `name`: "MyToken"
- `description`: "A decentralized exchange for trading ERC-20 tokens"
- `ens_name`: "mycontract.eth"

This enables contracts to self-identify with ENS names while maintaining consistent metadata across chains and through upgrades.

nxt3d committed:
Add Contract-Level Onchain Metadata ERC draft

- Defines standard for storing contract metadata onchain using ERC-7201 namespaced storage
- Provides predictable storage locations for cross-chain compatibility
- Extends ERC-7572 concept with onchain storage
- Includes interface, examples, and reference implementation

nxt3d committed:
Update created date to 2025-10-10

nxt3d committed:
Add pre-calculated storage location constant

nxt3d committed:
Switch from ERC-7201 to Diamond Storage pattern

- Replace ERC-7201 namespaced storage with Diamond Storage pattern
- Use simpler keccak256(namespace) calculation instead of complex ERC-7201 formula
- Update namespace to 'ercXXXX.contract.metadata.storage' for uniqueness

nxt3d committed:
Update contract metadata standard to reference ERC-8042

- Add requires: 8042 to header
- Update abstract to reference ERC-8042's Diamond Storage pattern
- Update security considerations to reference ERC-8042.

nxt3d committed:
Update discussions-to URL for contract metadata standard

abcoathup commented on ERCS/erc-draft_contract-metadata.md:
```diff
@@ -0,0 +1,136 @@
+---
+eip: TBD
```
```suggestion
eip: 8049
```
Assigning next sequential EIP/ERC/RIP number.
Numbers are assigned by editors & associates.

Please also update the filename.

nxt3d committed:
Update ERCS/erc-draft_contract-metadata.md

Co-authored-by: Andrew B Coathup <28278242+abcoathup@users.noreply.github.com>

nxt3d committed:
Update ERCS/erc-draft_contract-metadata.md

Co-authored-by: Andrew B Coathup <28278242+abcoathup@users.noreply.github.com>

nxt3d committed:
Rename file to erc-8049.md to match assigned ERC number

- Rename ERCS/erc-draft_contract-metadata.md to ERCS/erc-8049.md
- Reflects ERC editor's assignment of ERC-8049 to the contract metadata standard

nxt3d committed:
Make Diamond Storage optional in ERC-8049

- Remove 'requires: 8042' from header (now optional)
- Update title and abstract to reflect optional Diamond Storage
- Add 'Optional Diamond Storage' section with namespace 'erc8049.contract.metadata.storage'
- Add simple Basic Implementation without Diamond Storage
- Add Diamond Storage Implementation as separate example
- Update pragma to ^0.8.25 in both implementations
- Clarify benefits for cross-chain applications and upgradable contracts
- Update security considerations to only reference ERC-8042 for Diamond Storage implementations
- Both implementations tested and validated with comprehensive test suite

nxt3d committed:
Update ERC-8049: Add Optional Metadata Hooks section

- Add Optional Metadata Hooks section to ERC-8049
  - References ERC-hooks standard for full specification
  - Defines ContractMetadataHook selector (0xcae9bb7e)
  - Enables secure resolution from known contracts

- Remove forge-std submodule and .gitmodules
  - Clean up accidentally committed forge-std dependency

nxt3d committed:
ERC-8049: Enhance metadata key specification and reorganization

- Add new 'Optional Key Parameters' section explaining parameterized keys
- Specify TOON format (minimal agent-optimized format similar to YAML) for
  parameterized keys using colon and space separator
- Provide clear examples of recommended formats ('registration: 1') vs
  not recommended formats ('registration-1', 'registration:1', 'registration1')
- Improve documentation clarity and consistency in key/value pair handling

This enhancement provides clearer guidance for implementers on how to structure
metadata keys, especially when dealing with parameterized or variant metadata.

SamWilsn committed:
Update erc-8049.md

SamWilsn commented on ERCS/erc-8049.md:
```diff
@@ -0,0 +1,187 @@
+
+## Motivation
+
+ERC-7572 provides a standard for contract-level metadata via `contractURI()`, but it primarily focuses on offchain metadata storage. This ERC extends that concept by providing onchain storage. Contracts that need predictable storage locations for cross-chain compatibility or upgradable contracts can optionally use the Diamond Storage pattern from ERC-8042.
```
Your motivation is mostly just restating your abstract. Instead, this section should explain why the ecosystem needs this new standard. What use cases does it enable? Why is this particular proposal better than other (possibly hypothetical) competing proposals?

SamWilsn commented on ERCS/erc-8049.md:
```diff
@@ -0,0 +1,187 @@
+
+- `getContractMetadata(key)`: Returns the contract metadata value for the given key as bytes
+
+Contracts implementing this ERC MAY also expose a `setContractMetadata(string calldata key, bytes calldata value)` function to allow metadata updates, with write policy determined by the contract.
```
I'd recommend putting `setContractMetadata` into its own interface. That makes, for example, ERC-165 implementation easier.

nxt3d replied:
There are many ways that metadata can be set. I am unsure about this change.

SamWilsn commented on ERCS/erc-8049.md:
```diff
@@ -0,0 +1,187 @@
+
+### Optional Key Parameters
+
+Keys MAY include parameters to represent variations or instances of a metadata type. When keys include parameters, they SHOULD use TOON format (a minimal agent-optimized format similar to YAML, developed by Johann Schopplich) with a colon and space separator.
```
```suggestion
Keys MAY include parameters to represent variations or instances of a metadata type. When keys include parameters, they SHOULD use Token-Oriented Object Notation (TOON) format.
```

SamWilsn replied:
Mentioning TOON here is kinda an external reference. Generally we let this slide for JSON and other well-known formats, but I don't believe TOON qualifies as well-known yet.

We have two options here. Either we allow the TOON specification as an external origin (see [EIP-5757](https://eips.ethereum.org/EIPS/eip-5757)), or you just copy the spec (it's available under MIT) into your assets directory and link to it there. How would you like to proceed?

SamWilsn commented on ERCS/erc-8049.md:
```diff
@@ -0,0 +1,187 @@
+A contract can store basic information about itself:
+
+- Key: `"name"` → Value: `bytes("MyToken")`
+- Key: `"description"` → Value: `bytes("A decentralized exchange for trading ERC-20 tokens")`
```
If this is just an example, it'll be easier to remove the `ERC-20` from the string than to update the linting tool.

That said, this is technically a bug in the tool, so if you'd like to keep this, you may and I'll fix the tool.

SamWilsn commented on ERCS/erc-8049.md:
````diff
@@ -0,0 +1,187 @@
+### Diamond Storage Implementation
+
+```solidity
+// SPDX-License-Identifier: MIT
````
Same license comment.

SamWilsn reviewed (commented):
Looking pretty good so far! Just a few small changes; nothing major.

nxt3d committed:
Merge branch 'ethereum:master' into contract-metadata-erc

nxt3d committed:
Update ERC-8049 key parameter guidance

Simplify optional key parameters section to reference ERC-XXXX and remove SPDX license comments from code examples.

nxt3d commented on ERCS/erc-8049.md:
```diff
@@ -0,0 +1,187 @@
+---
+eip: 8049
+title: Contract-Level Onchain Metadata
+description: A standard for storing contract-level metadata onchain with optional Diamond Storage pattern for predictable storage locations.
```
ok nice!

nxt3d committed:
Apply suggestion from @SamWilsn

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>

nxt3d committed:
Improve ERC-8049 motivation

Clarify the ecosystem need for onchain contract metadata and note benefits of optional Diamond Storage for cross-chain proofs and upgrades.

nxt3d committed:
Remove ERC-8049 scope section

nxt3d committed:
Tweak ERC-8049 example description

nxt3d committed:
Add link to ERC-8119 in Optional Key Parameters section

nxt3d committed:
Update ERC references to [ERC-XXXX](./eip-xxxx.md) format in erc-8049

nxt3d committed:
Rephrase metadata hooks section in erc-8049

nxt3d commented on ERCS/erc-8049.md:
````diff
@@ -0,0 +1,187 @@
+### Basic Implementation
+
+```solidity
+// SPDX-License-Identifier: MIT
````
deleted the SPDX-License-Identifier

SamWilsn committed:
Merge branch 'master' into contract-metadata-erc

nxt3d committed:
Update ERC-8049 hooks section to match ERC-8121

Update hooks section to reference ERC-8121 and add use cases for KYC, PoP, and credentials.

SamWilsn reviewed (approved):
Fine for a draft, but do update your requires preamble field when you get the chance.

nxt3d committed:
Merge branch 'master' into contract-metadata-erc
